### PR TITLE
Add RoutingContext to Vert.x duplicated context local data so that it is accessible in the SecurityIdentityAugmentor

### LINKED ISSUE #28326: Add RoutingContext to SecurityIdentity for mTLS authentication
### Description

Hi,

I'm trying to access the current request in a `SecurityIdentityAugmentor`. Since injecting `RoutingContext` doesn't seem to work (I tried with and without `@ActivateRequestContext`), I'm wondering if `RoutingContext` can be passed in a `SecurityIdentity` attribute. I see it's already done in `OidcIdentityProvider` ([here](https://github.com/quarkusio/quarkus/blob/e19f65e535e639bdd542633471d1309a61364237/extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/OidcIdentityProvider.java#L225)) and `MpJwtValidator` ([here](https://github.com/quarkusio/quarkus/blob/5036e2f94f71f88f3e0a7415c7b2c519861a267c/extensions/smallrye-jwt/runtime/src/main/java/io/quarkus/smallrye/jwt/runtime/auth/MpJwtValidator.java#L83)). For my use case, I'd like to have the same behavior for mTLS authentication (`X509IdentityProvider`). If you think it's valuable, I can try to put an MR together.

Thanks,

lorenzo

### Implementation ideas

`MtlsAuthenticationMechanism` is [already adding](https://github.com/quarkusio/quarkus/blob/9e2c88643a8f3a1c5bff9e12dd5fdc1dfe4d628c/extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/security/MtlsAuthenticationMechanism.java#L60) the `RequestContext` to the `AuthenticationRequest` attributes, I think it only needs to be forwarded to the `SecurityIdentity` attributes in `X509IdentityProvider`:

```java
X509Certificate certificate = request.getCertificate().getCertificate();
RoutingContext routingContext = HttpSecurityUtils.getRoutingContextAttribute(request);
if (routingContext != null) {
    builder.addAttribute(RoutingContext.class.getName(), routingContext);
}
QuarkusSecurityIdentity.Builder builder = QuarkusSecurityIdentity.builder()
                .setPrincipal(certificate.getSubjectX500Principal())
                .addCredential(request.getCertificate())               
return Uni.createFrom().item(builder.build());
```

### LINKED ISSUE #37457: currentVertxRequest.getCurrent in a SecurityIdentityAugmentor became null since 3.2.9 with GraphQL
### Describe the bug

From 3.1.X to 3.2.8, the following SecurityIdentityAugmentor implementation:

```java
@ApplicationScoped
public class SecurityAugmentor implements SecurityIdentityAugmentor {

    @Inject
    private OidcClient oidcClient;

    @Inject
    private OidcIdentityProvider oidcProvider;

    @Inject
    private CurrentVertxRequest currentVertxRequest;

    @Override
    public Uni<SecurityIdentity> augment(SecurityIdentity identity, AuthenticationRequestContext context) {
        if (identity.isAnonymous()) {
            // Not authenticated calls come here, grant "Guest" account from configured Service Account
            return buildGuestIdentity(identity, context);
        }
        return context.runBlocking(QuarkusSecurityIdentity.builder(identity)::build);
    }

    @ActivateRequestContext
    public Uni<SecurityIdentity> buildGuestIdentity(SecurityIdentity identity, AuthenticationRequestContext context) {
        return oidcClient.getTokens(Map.of()).chain(tokens -> {

            RefreshToken rt = new RefreshToken(tokens.getRefreshToken());
            AccessTokenCredential atc = new AccessTokenCredential(tokens.getAccessToken(), rt);

            // Prepare an authentication request to call the OIDC identity provider, or else the current
            // request will be seen has not authenticated
            var tar = new TokenAuthenticationRequest(atc);
            // Save the RountingContext (URL ...) of the current request in the authentication request
            if (currentVertxRequest != null && currentVertxRequest.getCurrent() != null) {
                HttpSecurityUtils.setRoutingContextAttribute(tar, currentVertxRequest.getCurrent());
                return oidcProvider.authenticate(tar, context);
            } else {
                // A query is received on a not functional path
                // Either health check or flood, keep the incoming identity (will result in a 401)
                return Uni.createFrom().item(identity);
            }
            // Let call again Keycloak to get proper authenticated context
        }).onItem().transform(s -> build(s).get());
    }
```

was able to retrieve a `currentVertxRequest.getCurrent()` value upon unauthenticated requests on any path of my application.
Then the class was granting a limited right service account authentication to simulate kind of a not authenticated Guest access to all endpoints.
   
From 3.2.9, the returned current Vertx context is *null*.  
Trying to workaround with RoutingContext injection leads to the following exception (in both POST GraphQL or GET rest http requests):
```
2023-12-01 15:31:29 [vert.x-eventloop-thread-3] ERROR io.quarkus.vertx.http.runtime.QuarkusErrorHandler - HTTP Request to /graphql failed, error id: da5cd711-ea78-46de-8a53-34b1c1a95605-1
jakarta.enterprise.inject.IllegalProductException: Normal scoped producer method may not return null: io.quarkus.vertx.http.runtime.CurrentVertxRequest.getCurrent()
        at io.quarkus.vertx.http.runtime.CurrentVertxRequest_ProducerMethod_getCurrent_6dc23d16d53ba5c34e1e7b6f54290fd7b9aebd76_Bean.doCreate(Unknown Source)
        at io.quarkus.vertx.http.runtime.CurrentVertxRequest_ProducerMethod_getCurrent_6dc23d16d53ba5c34e1e7b6f54290fd7b9aebd76_Bean.create(Unknown Source)
        at io.quarkus.vertx.http.runtime.CurrentVertxRequest_ProducerMethod_getCurrent_6dc23d16d53ba5c34e1e7b6f54290fd7b9aebd76_Bean.create(Unknown Source)
        at io.quarkus.arc.impl.RequestContext.getIfActive(RequestContext.java:74)
        at io.quarkus.arc.impl.ClientProxies.getDelegate(ClientProxies.java:30)
        at io.vertx.ext.web.CurrentVertxRequest_ProducerMethod_getCurrent_6dc23d16d53ba5c34e1e7b6f54290fd7b9aebd76_ClientProxy.arc$delegate(Unknown Source)
        at io.vertx.ext.web.CurrentVertxRequest_ProducerMethod_getCurrent_6dc23d16d53ba5c34e1e7b6f54290fd7b9aebd76_ClientProxy.put(Unknown Source)
        at io.quarkus.oidc.runtime.OidcIdentityProvider.authenticate(OidcIdentityProvider.java:70)
        at io.quarkus.oidc.runtime.OidcIdentityProvider_ClientProxy.authenticate(Unknown Source)
        at <<<< The security Augmentor on line oidcProvider.authenticate >>>>
        at io.smallrye.context.impl.wrappers.SlowContextualFunction.apply(SlowContextualFunction.java:21)
        at io.smallrye.mutiny.operators.uni.UniOnItemTransformToUni$UniOnItemTransformToUniProcessor.performInnerSubscription(UniOnItemTransformToUni.java:68)
        at io.smallrye.mutiny.operators.uni.UniOnItemTransformToUni$UniOnItemTransformToUniProcessor.onItem(UniOnItemTransformToUni.java:57)
        at io.smallrye.mutiny.operators.uni.UniOnItemTransform$UniOnItemTransformProcessor.onItem(UniOnItemTransform.java:43)
        at io.smallrye.mutiny.operators.uni.UniOperatorProcessor.onItem(UniOperatorProcessor.java:47)
        at io.smallrye.mutiny.operators.uni.UniOperatorProcessor.onItem(UniOperatorProcessor.java:47)
        at io.smallrye.mutiny.vertx.AsyncResultUni.lambda$subscribe$1(AsyncResultUni.java:35)
        at io.smallrye.mutiny.vertx.DelegatingHandler.handle(DelegatingHandler.java:25)
        at io.vertx.ext.web.client.impl.HttpContext.handleDispatchResponse(HttpContext.java:397)
        at io.vertx.ext.web.client.impl.HttpContext.execute(HttpContext.java:384)
        at io.vertx.ext.web.client.impl.HttpContext.next(HttpContext.java:362)
        at io.vertx.ext.web.client.impl.HttpContext.fire(HttpContext.java:329)
        at io.vertx.ext.web.client.impl.HttpContext.dispatchResponse(HttpContext.java:291)
        at io.vertx.ext.web.client.impl.HttpContext.lambda$null$7(HttpContext.java:507)
        at io.vertx.core.impl.ContextInternal.dispatch(ContextInternal.java:277)
        at io.vertx.core.impl.ContextInternal.dispatch(ContextInternal.java:259)
        at io.vertx.core.impl.EventLoopContext.lambda$runOnContext$0(EventLoopContext.java:43)
        at io.netty.util.concurrent.AbstractEventExecutor.runTask(AbstractEventExecutor.java:173)
        at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:166)
        at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:470)
        at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:569)
        at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997)
        at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        at java.base/java.lang.Thread.run(Thread.java:833)
```

### Expected behavior

`currentVertxRequest.getCurrent()` still returns something when an Http query (REST or GraphQL) has been received and routed to the SecurityAugmentor.

### Actual behavior

`currentVertxRequest.getCurrent()` returns null.
Injection of `RutingContext` leads to an exception.

### How to Reproduce?

Steps to reproduce:
1. Add the proposed security Augmentor to any Quarkus application that handle OIDC authentication
2. Launch a query to an authenticated path of the application without token

### Output of `uname -a` or `ver`

Windows

### Output of `java -version`

java version "17" 2021-09-14 LTS Java(TM) SE Runtime Environment (build 17+35-LTS-2724) Java HotSpot(TM) 64-Bit Server VM (build 17+35-LTS-2724, mixed mode, sharing)

### Quarkus version or git rev

3.2.9.Final

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Apache Maven 3.8.2

### Additional information

The 3.2.9 release note includes some backports (from 3.7) related to RoutingContext, GraphQL (and Vertx), mainly for WebSocket, but I didn't see any direct correlation from those commits that could have lead to this current change of behavior.
Neither did I found any other way to retrieve the RoutingContext in the Tests that were added.

### PR DESCRIPTION
**Pull request by michalvavrik**
- fixes #37457
- fixes #28326
- we have seen other requests to make `RoutingContext` accessible from the `SecurityIdentityAugmentor` like in #16861

### CODE REVIEW

**Comment by michalvavrik** [comment]
I've seen both of these CI failures in my other PR.s

File: extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/VertxHttpRecorder.java
Comment (sberyozkin) [inline extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/VertxHttpRecorder.java:493]: Hi @michalvavrik This PR is easy to understand and is totally fine, but I'd like to hear from @franz1981 and @cescoffier what do they think about the possible performance impact, cheers

Comment (michalvavrik) [reply extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/VertxHttpRecorder.java:493]: Sure, let's gather feedback.

FYI I can easily avoid creating handler by putting in for example here:

https://github.com/quarkusio/quarkus/blob/cdbb8e159aab00719a4fbef8fd32150eb304fea8/extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/security/HttpSecurityRecorder.java#L244

but I thought having `RoutingContext` in duplicated context could be potentially beneficial to others as well, so I added it right after an HTTP request processing started. If this view is not shared, I will move it.

Thanks

Comment (michalvavrik) [reply extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/VertxHttpRecorder.java:493]: After some thinking, avoiding extra handler is probably a good idea @sberyozkin , I'll address it.

Comment (michalvavrik) [reply extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/VertxHttpRecorder.java]: Done.

File: extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/security/HttpSecurityUtils.java
Comment (franz1981) [inline extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/security/HttpSecurityUtils.java:33]: This will blindly set it always, even if present already: this is what we need? or we need avoiding setting it again if already present, and failing if what's present is different (or any other behaviour we want to enforce).

Sadly `ctx::putLocal` is  a `ConcurrentHashMap::put` which can become way cheaper if it doesn't always happen (unless is the semantic we requires, or we know that is always transitioning the value from being absent to the new value).

File: extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/security/HttpSecurityUtils.java
Comment (franz1981) [inline extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/security/HttpSecurityUtils.java:32]: it's just ignored? we don't want to throw an exception here?

File: extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/security/HttpSecurityRecorder.java
Comment (franz1981) [inline extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/security/HttpSecurityRecorder.java:242]: Just out of curiosity: why using `Class::getName`?
If we have a serie of possible keys somewhere, I would put something there, with a specific name unrelated to the class's name, which can still bother the JVM to init it  (and cache it , see https://github.com/openjdk/jdk/blob/f55381950266088cc0284754b16663675867ac87/src/java.base/share/classes/java/lang/Class.java#L988C43-L988C43).

File: extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/security/HttpSecurityUtils.java
Comment (michalvavrik) [reply extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/security/HttpSecurityUtils.java:32]: That's a good question. Situation is that I expect this always to be true at this moment because in `VertxHttpRecorder` the duplicated context is enforced, however I won't risk an exception and I think it is safer to always check. 

So no, I acknowledge there can be extensions that for some reason will try to use root context and then feature that relies on duplicated context won't work. Please note that documentation explicitly mention this feature depends on duplicated context so users can check https://quarkus.io/guides/duplicated-context#context-local-data and see if they can use it or not.

Comment (franz1981) [reply extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/security/HttpSecurityUtils.java:32]: If `setRoutingContextAttribute` is exposed to external users, to me, would be better to check it and fail/report somehow what's going on.
if the users are "internals" - still other known extensions we have control of and test coverage, you can just use an `assert` which should be enabled on tests IIRC (@geoand can confirm, I'm not sure...and TBH not everyone love asserts, so it's a matter of personal preferences...)

File: extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/security/HttpSecurityRecorder.java
Comment (michalvavrik) [reply extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/security/HttpSecurityRecorder.java:242]: I just wanted to follow pattern that users already know in Quarkus OIDC (for example) 

https://github.com/quarkusio/quarkus/blob/447067819a58a10841375f41c12c36fb0eef5dbf/extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/OidcUtils.java#L349

I'm fine with introducing String constant, but there can be some users already doing `RoutingContext.class.getName` elsewhere.

Comment (franz1981) [reply extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/security/HttpSecurityRecorder.java:242]: This is a good point; so let's still with what's the expected usage than :+1:

File: extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/security/HttpSecurityUtils.java
Comment (michalvavrik) [reply extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/security/HttpSecurityUtils.java:33]: I think relation of RoutingContext to duplicated context is 1:N (but at this point it will most likely be 1:1, just in case someone create a new one later?) so it will always be the same. And it is not there otherwise we would not bother to add it there. I can't presume if situation can change in the future.

Sorry I don't have a good answer for this, but I appreciate your input. I'd like to always add it there because right now we are talking augmentors, but users can't access it anywhere during authentication and authorization (that is simplified - list of cases is also complicated)

File: extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/security/HttpSecurityUtils.java
Comment (michalvavrik) [reply extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/security/HttpSecurityUtils.java:32]: But test coverage won't help here, we can't control if some user code or maybe Quarkiverse extension will mark https://github.com/quarkusio/quarkus/blob/447067819a58a10841375f41c12c36fb0eef5dbf/extensions/vertx/runtime/src/main/java/io/quarkus/vertx/core/runtime/context/VertxContextSafetyToggle.java#L101 as unsafe or will enforce root context.

This feature makes sense because duplicated context should be used in most cases and only some rare cases will not have `RoutingContext` available.

Is this maybe problem of documentation, do you think I didn't highlight it enough that this feature depends on duplicated context? Please let me know.

File: extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/security/HttpSecurityUtils.java
Comment (michalvavrik) [reply extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/security/HttpSecurityUtils.java:33]: Quick check @franz1981  - do you think it makes sense to always run `get != null` before adding it even though it won't be the case unless something changes in the future? I can add it.

Comment (franz1981) [reply extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/security/HttpSecurityUtils.java:33]: To me it's fine as it is as long as the semantic for who use it is clear and doc somewhere :+1:

Comment (michalvavrik) [reply extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/security/HttpSecurityUtils.java:33]: Well I thought that would be answer on not executing unnecessary `ConcurrentHashMap::put`, but as said, unless someone starts to add it, this code runs pretty early so it won't be there.

Comment (michalvavrik) [reply extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/security/HttpSecurityUtils.java:33]: Okay, so I have added extra condition to address your comment.

File: extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/security/HttpSecurityUtils.java
Comment (michalvavrik) [reply extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/security/HttpSecurityUtils.java]: > If setRoutingContextAttribute is exposed to external users, to me, would be better to check it and fail/report somehow what's going on.

Okay, I've added extra logging for this situation, but please not this is best effort feature which as far as my knowledge of duplicated context goes, will be available for most users. Does it look better @franz1981 ?

**Comment by sberyozkin** [comment]
Hey Francesco, @franz1981,  thanks for the feedback to Michal's PR, and apologies I used the first 5 letters of your github alias when asking for your feedback in the now resolved comment, I switched off for 1 sec thinking it was the right name :-)

**Comment by michalvavrik** [comment]
`HttpSecurityPolicySecurityEventTest.testRolesPolicy` I didn't wrote correctly I can see now, need little tweak on condition to unmake it flaky, but it's not related to this PR.

`VertxIOHangTestCase` seems flaky, can't think of a way it could be related to this PR.

**Comment by michalvavrik** [comment]
I'll look if I can fix HTTP Sec policy in this PR, it should be tiny fix.

**Comment by gsmet** [comment]
We need to make sure that this doesn't have a cost that is too important either from an allocation standpoint or a memory usage standpoint but I see discussions are already going on.

Let's also make sure that we have @cescoffier 's blessing before merging this.

**Comment by franz1981** [comment]
I would like @geoand as well, cause I have no idea if this is a fixed cost we always expect to pay or is just is a specific feature is enabled

**Comment by geoand** [comment]
I need to go through this, so far I was only paying very loose attention

**Comment by michalvavrik** [comment]
It should be noted because `RoutingContext` is not available in general, we keep adding it to authentication request attributes and `SecurityIdentity` attributes.

>  fixed cost we always expect to pay or is just is a specific feature is enabled.

It is always, sure we can have some detection of augmentors if that is required, which will make this feature related directly only to augmentors and not making the context generally available during auth, but I'd like to first understand why there are so worries about one instance stored in the map. Can you help me to understand what is the cost you are referring to? Can you compare it with other Quarkus code when anything put in the map is investigated this way? Thank you

**Comment by franz1981** [comment]
> but I'd like to first understand why there are so worries about one instance stored in the map. Can you help me to understand what is the cost you are referring to?

Context::putLocal is performing this under the hood: https://github.com/eclipse-vertx/vert.x/blob/296a325c99bf77a56de1a49dd8d32f400cd78fc1/src/main/java/io/vertx/core/impl/DuplicatedContext.java#L123's `put` by using synchronized to retrieve the CHM field.
Sadly `synchronized`, not that biased locking is gone, are performed using 2 atomic volatile ops to enter/exit the monitor (not sure about the exit, but it should be like that) AND `ConcurrentHashMap::put`, under the hood require to both allocate an `entry` (to store the key and the value) AND perform at least an atomic op the same (see https://github.com/openjdk/jdk/blob/master/src/java.base/share/classes/java/util/concurrent/ConcurrentHashMap.java#L1019).

The problem with atomic operations is while performed in the hot path and near to each others (without meaningful work in-between) can be quite costy because they forcibly "await" the so called store buffer of the cpu cache they run to be fully drained: the operation per-se isn't costy, but prevent the cpu to fully use its instruction-level-parallelism, causing cpu stalls which the cpu could have used to perform entire methods, instead, or just meaningful work. 

More info about it at https://github.com/quarkusio/quarkus/pull/30024, including another problem due to conditional-card-marking (explained in the issue).

> Can you compare it with other Quarkus code when anything put in the map is investigated this way? Thank you

The PR I've shown should report some of the analysis performed to evaluate the cost, although it refer, as usual, to synthetic use cases, mostly CPU bound: in the "real" world this cost maybe won't be as bad, but the inherent overhead it brings, if made wrong, could still bite us for some real users and 100% in public benchmarks, which we cannot ignore.

**Comment by sberyozkin** [comment]
Hi All, and @franz1981 @geoand @cescoffier @gsmet 

I'd like to clarify as well, in the original code this PR had there was a new handler created to set a RoutingContext attribute property on the duplicated context, but not it has gone. 
The only action which happens now is that this property is added to the context at
https://github.com/quarkusio/quarkus/pull/37795/files#diff-931a1532759e6852f492ef461181032a1249b0a5eb4d3eb594fc8f6350814be5R37

and this is it, there are just about 2 very small source updates to make it happen - IMHO now that the extra handler has gone this PR has no any performance related impact.

The reason this is done is to workaround limitations of some security API, such as `SecurityIdentity`, since users often need access to the `RoutingContext` in security identity augmentors.

**Review by sberyozkin — APPROVED** [review]
LGTM, but awaiting for one more approval

**Comment by sberyozkin** [comment]
>IMHO now that the extra handler has gone this PR has no any performance related impact.

and then I saw the analysis from Francesco @franz1981 :-).

Michal @michalvavrik Can we just do it conditionally, HTTP security authentication mechanisms set this property anyway, so may be it should only be done when the duplicated contexts is accessed there  ?

**Comment by michalvavrik** [comment]
Thanks very much @sberyozkin , but I simply can't have answer to @franz1981 . I love performance, I'm just not sure whether every other Quarkus addition is analyzed with such a scrutiny in regards to performance. If that is going to be new standard, I think it's fantastic and I'm really happy.


> Michal @michalvavrik Can we just do it conditionally, HTTP security authentication mechanisms set this property anyway, so may be it should only be done when the duplicated contexts is accessed there ?

We don't work with duplicated context I think. @sberyozkin what about we just change `SecurityIdentityAugmentor` API and pass down to augmentor `io.quarkus.security.identity.request.AuthenticationRequest#getAttributes`?

**Comment by michalvavrik** [comment]
So I think I'm going to close this PR it is only related to duplicated context local data and that was rejected. I'll open PR In Quarkus Security project with proposal.

**Comment by michalvavrik** [comment]
thanks @franz1981

**Comment by sberyozkin** [comment]
@michalvavrik 

>what about we just change SecurityIdentityAugmentor API 

This will be a significant breaking change.

What about https://github.com/quarkusio/quarkus/blob/main/extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/AbstractOidcAuthenticationMechanism.java#L41 ?

Can we use a more generic property enabling such a propagation ?

**Comment by sberyozkin** [comment]
Hi @michalvavrik 

>I'm just not sure whether every other Quarkus addition is analyzed with such a scrutiny in regards to performance

This is probably the hottest request execution path in Quarkus so I guess we have to accept that only the most critical updates can be allowed to add something to the overall processing time here, given all the public attention to the performance of Quarkus. We can try limit it at the security authentication path only

**Comment by geoand** [comment]
Was this closed by accident?

**Comment by geoand** [comment]
By the way, this looks good to me.

**Comment by sberyozkin** [comment]
@michalvavrik Maybe we can simply consider doing what OIDC does and update every other authentication mechanism and identity provider we ship to have it added a security identity attribute, won't cover custom identity providers though but  we will just doc how custom identity providers can do it, that will be simple and I'm nearly 100% sure, have no negative performance impact.

@geoand thanks, I believe Michal got concerned he could not rework it to address the performance concerns, so he closed it, though maybe a little bit too fast

**Comment by geoand** [comment]
As for the performance concerns, we would need to measurements. There is just know way to guess with this kind of thing.

**Comment by michalvavrik** [comment]
> @michalvavrik Maybe we can simply consider doing what OIDC does and update every other authentication mechanism and identity provider we ship to have it added a security identity attribute, 

No, this won't help because usecase were also for anonymous identity and re-creating anonymous identity wouldn't be the best for performance we can do.

> won't cover custom identity providers though but we will just doc how custom identity providers can do it, that will be simple and I'm nearly 100% sure, have no negative performance impact.

I agree, custom identity providers will have to do it themselves, but I suggest to pass down the map where we already have the `RoutingContext` as I suggested earlier. I'll open separate PR in Quarkus Security where we can discuss it (and you can come with alternative proposals). Too many people that might not be interested in it are subscribed here.

> @geoand thanks, I believe Michal got concerned he could not rework it to address the performance concerns, so he closed it, though maybe a little bit too fast

I just accepted @franz1981 arguments, I opened this PR because I thought it's not a big deal to add one more item to contextual local data, but there are ways to completely avoid it now that I understand problem. Sorry to get so many people involved, but this was very informative, thanks.

**Comment by michalvavrik** [comment]
Let's continue here https://github.com/quarkusio/quarkus-security/pull/40 @sberyozkin .

**Pull request closed without merging**
